sql/000002.sql

```diff
@@ -0,0 +1,183 @@
+      values(
+        identifier,
+        coalesce(payload, '{}'),
+        coalesce(queue_name, public.gen_random_uuid()::text),
+        coalesce(run_at, now()),
+        coalesce(max_attempts, 25),
+        job_key
+      )
+      on conflict (key) do update set
+        -- update job details if provided, otherwise maintain existing value
+        -- note we can't use excluded here, or the value will always be updated,
+        -- due to the coalesce statements in values ( ... ) above
+        task_identifier=coalesce(add_job.identifier, jobs.task_identifier),
+        payload=coalesce(add_job.payload, jobs.payload),
+        queue_name=coalesce(add_job.queue_name, jobs.queue_name),
+        max_attempts=coalesce(add_job.max_attempts, jobs.max_attempts),
+
+        -- update run_at if argument is provided. If not, and there has been an
+        -- error, assume run_at reflects the error retry backoff, so reset it.
+        -- If no errors, maintain existing value by default
+        run_at=coalesce(
+          add_job.run_at,
+          (case when jobs.attempts > 0
+            then now()
+            else jobs.run_at
+          end)
+        ),
+
+        -- always reset error/retry state
+        attempts=0,
+        last_error=null
+      where jobs.locked_at is null
+      returning *
+      into v_job;
+
+    -- if the returned id is null, assume job is already locked for processing
+    -- and couldn't be updated
+    if v_job.id is null then
+      -- remove existing key to allow a new one to be inserted, and prevent any
+      -- subsequent retries by bumping attempts to the max allowed
+      update graphile_worker.jobs set
+        key = null,
+        attempts = jobs.max_attempts
+      where key = job_key;
+
+      -- insert the new job. Assume no conflicts due to the update above
+      insert into graphile_worker.jobs(task_identifier, payload, queue_name, run_at, max_attempts, key)
+        values(
+          identifier,
+          coalesce(payload, '{}'),
+          coalesce(queue_name, public.gen_random_uuid()::text),
+          coalesce(run_at, now()),
+          coalesce(max_attempts, 25),
+          job_key
+        )
+        returning *
+        into v_job;
+    end if;
+  end if;
+  return v_job;
+end;
+$$ language plpgsql;
+
+
+--- implement new remove_job function
+
+create function graphile_worker.remove_job(
+  job_key text
+) returns graphile_worker.jobs as $$
+  delete from graphile_worker.jobs
+    where key = job_key
+    and locked_at is null
+  returning *;
+$$ language sql;
+
+
+-- Update other functions to handle locked_at denormalisation
+
+create or replace function graphile_worker.get_job(worker_id text, task_identifiers text[] = null, job_expiry interval = interval '4 hours') returns graphile_worker.jobs as $$
+declare
+  v_job_id bigint;
+  v_queue_name text;
+  v_default_job_max_attempts text = '25';
+  v_row graphile_worker.jobs;
+  v_now timestamptz = now();
```

Comment:
Technically `now()` will never change within a transaction, so this is unnecessary. I think it makes things more explicit though, so we'll keep it.

Reply:
Could you elaborate on this? Which part is unnecessary

Reply:
You refactored the code to use the `v_now` variable rather than using the `now()` call directly in various different locations within the function (you can look [at the old code](https://github.com/graphile/worker/blob/52b776c1d83ffdbbd29cf053c80bfcabe29b0fb3/sql/000001.sql#L84-L124) to see us calling `now()` in a couple different statements in the function without a `v_now` variable).

In case you're not familiar, here's where `NOW()` is documented:

https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-CURRENT

Of relevance (emphasis mine):

> These SQL-standard functions all return values based on **the start time of the current transaction**:

and

> Since these functions return the start time of the current transaction, **their values do not change during the transaction**. 

Functions always run within a single transaction, so all `now()` calls in the function will have the same value and there's no need to factor them out to a `v_now` variable. That said, I don't think this does any harm and it makes the code clearer for people not familiar with this concept so I'm inclined to keep it.

Reply:
Ah right, I see. Thanks for the explanation :)